### PR TITLE
Async jobs add endtime

### PR DESCRIPTION
**Pull request by ernjvr**
## Description
Overview:
There is currently no functional mechanism that captures or persists the end time of when an asynchronous job has finished. As a result, users are not able to do any reporting about the duration of various asynchronous jobs in Cloudstack.
* Data Layer changes:
The time stamp of when an asynchronous job has finished should be persisted to the 'removed' column of the async_job database table. This is an existing column which is not being used for any other functionality.
* QueryAsyncJobResult API changes:
The queryAsyncJobResult API should now include an 'end_time' response tag that will display the value of the 'removed' database column.
* Management Server process changes:
When a Cloudstack management server starts up, it finds in the database all the asynchronous jobs it is the owner of, whose statuses are in an 'in_progress' state and updates them to a 'failed' status. At the same time, it should now also mark them as 'removed' by updating the 'removed' field with the current timestamp.
* Garbage collection changes:
Currently, the garbage collector uses the 'created' database column to find in the database completed asynchronous job records that are older than the 'job.expire.minutes' global setting's timestamp and, then, deletes these records. It should no longer use the 'created' column to do this. It should use the 'removed' column instead.

## Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)

## How Has This Been Tested?

* Start a virtual machine to initiate an asynchronous job. Query the cloud.async_job table to find the record that was created in response to the 'start virtual machine' job. When the job finishes, verify that the 'removed' column has been correctly updated with the end timestamp of the relevant record.
* Using cloudmonkey, make an API call to queryAsyncJobResult using the uuid as the jobid and verify that the 'endtime' field returns the correct timestamp value of the relevant database record's 'removed' column.
* Verify that the garbage collector deletes the correct expired asynchronous job records from the cloud.async_job table.
* Hypervisor: KVM

## Checklist:

- [x] I have read the [CONTRIBUTING](https://github.com/apache/cloudstack/blob/master/CONTRIBUTING.md) document.
- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
Testing
- [x] I have added tests to cover my changes.
- [ ] All relevant new and existing integration tests have passed.
- [ ] A full integration testsuite with all test that can run on my environment has passed.

### CODE REVIEW

File: api/src/main/java/org/apache/cloudstack/api/response/AsyncJobResponse.java [hunk 1]
```diff
@@ -75,6 +75,10 @@
     @Param(description = "  the created date of the job")
     private Date created;
 
+    @SerializedName(ApiConstants.END_TIME)
```
Comment (yadvr) [inline api/src/main/java/org/apache/cloudstack/api/response/AsyncJobResponse.java]: This could be finished or completed?

File: framework/jobs/src/main/java/org/apache/cloudstack/framework/jobs/dao/AsyncJobDaoImpl.java [hunk 2]
```diff
@@ -71,7 +72,7 @@ public AsyncJobDaoImpl() {
         expiringUnfinishedAsyncJobSearch.done();
 
         expiringCompletedAsyncJobSearch = createSearchBuilder();
-        expiringCompletedAsyncJobSearch.and("created", expiringCompletedAsyncJobSearch.entity().getCreated(), SearchCriteria.Op.LTEQ);
```
Comment (yadvr) [inline framework/jobs/src/main/java/org/apache/cloudstack/framework/jobs/dao/AsyncJobDaoImpl.java:74]: This change is okay only if the 'created' it not used in any of the methods to query the result.

File: framework/jobs/src/main/java/org/apache/cloudstack/framework/jobs/dao/AsyncJobDaoImpl.java [hunk 3]
```diff
@@ -168,11 +169,11 @@ public void cleanupPseduoJobs(long msid) {
     }
 
     @Override
-    public List<AsyncJobVO> getExpiredCompletedJobs(Date cutTime, int limit) {
-        SearchCriteria<AsyncJobVO> sc = expiringCompletedAsyncJobSearch.create();
-        sc.setParameters("created", cutTime);
+    public List<AsyncJobVO> getExpiredCompletedJobs(final Date cutTime, final int limit) {
+        final SearchCriteria<AsyncJobVO> sc = expiringCompletedAsyncJobSearch.create();
```
Comment (yadvr) [inline framework/jobs/src/main/java/org/apache/cloudstack/framework/jobs/dao/AsyncJobDaoImpl.java:173]: We'll need to discuss this bit over our syncup call when you're back.

File: framework/jobs/src/main/java/org/apache/cloudstack/framework/jobs/impl/AsyncJobManagerImpl.java [hunk 4]
```diff
@@ -302,14 +302,16 @@ public void completeAsyncJob(final long jobId, final Status jobStatus, final int
                     job.setResult(null);
                 }
 
-                job.setLastUpdated(DateUtil.currentGMTTime());
+                final Date currentGMTTime = DateUtil.currentGMTTime();
+                job.setLastUpdated(currentGMTTime);
+                job.setRemoved(currentGMTTime);
```
Comment (yadvr) [inline framework/jobs/src/main/java/org/apache/cloudstack/framework/jobs/impl/AsyncJobManagerImpl.java:307]: LGTM.

File: framework/jobs/src/main/java/org/apache/cloudstack/framework/jobs/impl/AsyncJobManagerImpl.java [hunk 5]
```diff
@@ -445,8 +447,8 @@ public void syncAsyncJobExecution(AsyncJob job, String syncObjType, long syncObj
     }
 
     @Override
-    public AsyncJob queryJob(long jobId, boolean updatePollTime) {
-        AsyncJobVO job = _jobDao.findById(jobId);
+    public AsyncJob queryJob(final long jobId, final boolean updatePollTime) {
+        final AsyncJobVO job = _jobDao.findByIdIncludingRemoved(jobId);
```
Comment (yadvr) [inline framework/jobs/src/main/java/org/apache/cloudstack/framework/jobs/impl/AsyncJobManagerImpl.java:451]: This may have some API response changes, wrt who uses this method and if it can cause any regression (as the method will return a removed job).

**Review by yadvr — CHANGES_REQUESTED** [review]
Some changes request, overall LGTM good work. I've not tested/checked your marvin test cases though.

File: test/integration/smoke/test_privategw_acl.py [hunk 6]
```diff
@@ -720,7 +720,7 @@ def check_pvt_gw_connectivity(self, virtual_machine, public_ip, vms_ips):
         succeeded_pings = 0
         minimum_vms_to_pass = 2
         for vm_ip in vms_ips:
-            ssh_command = "ping -c 5 %s" % vm_ip
+            ssh_command = "ping -c 10 %s" % vm_ip
```
Comment (borisstoyanov) [inline test/integration/smoke/test_privategw_acl.py]: @ernjvr is this related change? if it's a bugfix can we have it in a separate PR?

**Review by borisstoyanov — COMMENTED** [review]
@ernjvr I see lots of other commits in your PR, can you pull and rebase your code against master and resubmit the pull request

File: api/src/main/java/org/apache/cloudstack/api/response/AsyncJobResponse.java [hunk 7]
```diff
@@ -75,6 +75,10 @@
     @Param(description = "  the created date of the job")
     private Date created;
 
+    @SerializedName(ApiConstants.COMPLETED)
+    @Param(description = "  the removed date of the job")
```
Comment (yadvr) [inline api/src/main/java/org/apache/cloudstack/api/response/AsyncJobResponse.java]: Change the description and variable to completed?

File: framework/jobs/src/main/java/org/apache/cloudstack/framework/jobs/dao/AsyncJobDao.java [hunk 8]
```diff
@@ -24,6 +24,10 @@
 import com.cloud.utils.db.GenericDao;
 
 public interface AsyncJobDao extends GenericDao<AsyncJobVO, Long> {
+
+    @Override
+    AsyncJobVO findById(Long id);
```
Comment (yadvr) [inline framework/jobs/src/main/java/org/apache/cloudstack/framework/jobs/dao/AsyncJobDao.java]: Why is this necessary?

**Review by yadvr — CHANGES_REQUESTED** [review]

File: framework/jobs/src/main/java/org/apache/cloudstack/framework/jobs/dao/AsyncJobDaoImpl.java [hunk 9]
```diff
@@ -95,6 +100,26 @@ public AsyncJobDaoImpl() {
 
     }
 
+    @Override
+    public AsyncJobVO findById(Long id) {
```
Comment (yadvr) [inline framework/jobs/src/main/java/org/apache/cloudstack/framework/jobs/dao/AsyncJobDaoImpl.java]: The findByIdIncluding removed previous usage was okay.

File: framework/jobs/src/main/java/org/apache/cloudstack/framework/jobs/dao/AsyncJobDaoImpl.java (hunk 3 above)
Comment (yadvr) [reply framework/jobs/src/main/java/org/apache/cloudstack/framework/jobs/dao/AsyncJobDaoImpl.java:173]: Suggestion: during the upgrade 4.11.1->4.12.0 -> set all the removed field of the async_jobs -> `NOW()`?

File: test/integration/smoke/test_async_job.py [hunk 10]
```diff
@@ -0,0 +1,160 @@
...
+        response = self.virtual_machine.getState(
+            self.apiclient,
+            VirtualMachine.RUNNING)
+        self.assertEqual(response[0], PASS, response[1])
+
+        cmd = queryAsyncJobResult.queryAsyncJobResultCmd()
+        cmd.jobid = self.virtual_machine.jobid
+        cmd_response = self.apiclient.queryAsyncJobResult(cmd)
+        endtime = cmd_response.endtime
+        self.assertIsNotNone(endtime, "Expected 'endtime' field of queryAsyncJobResult to have a timestamp of when the job finished.")
```
Comment (yadvr) [inline test/integration/smoke/test_async_job.py]: cmd_response -> get the complete time, then use the `dbclient` -> async_jobs table -> job id and verify that completed matches what you got in the API response.

File: test/integration/smoke/test_async_job.py [hunk 11]
```diff
@@ -0,0 +1,160 @@
...
+
+    @attr(
+        tags=[
+            "advanced",
+            "eip",
+            "advancedns",
+            "basic",
+            "sg"],
+        required_hardware="true")
+    def test_async_job_table_values(self):
```
Comment (yadvr) [inline test/integration/smoke/test_async_job.py]: Squash the above two test cases into one, asset equal for the job id's removed/completed time.

File: test/integration/smoke/test_async_job.py [hunk 12]
```diff
@@ -0,0 +1,160 @@
...
+        """
+        Test async job db table for valid values
+        """
```
Comment (yadvr) [inline test/integration/smoke/test_async_job.py]: Add another failure test case. Destroy something twice, etc. job status -> failed (int 2?).

File: test/integration/smoke/test_async_job.py (hunk 10 above)
Comment (yadvr) [reply test/integration/smoke/test_async_job.py]: In the API, jobstatus -> success/failure -> check that against db as well.

**Review by yadvr — CHANGES_REQUESTED** [review]
Please see comments. For integration tests, make use of both the API respones and db client/query result to verify both the job status and the removed time.

File: framework/jobs/src/main/java/org/apache/cloudstack/framework/jobs/dao/AsyncJobDao.java (hunk 8 above)
Comment (ernjvr) [reply framework/jobs/src/main/java/org/apache/cloudstack/framework/jobs/dao/AsyncJobDao.java]: Because it was referenced by interface and not implementation. But as discussed, I will remove the override and use findByIdIncludingRemoved.

File: test/integration/smoke/test_async_job.py (hunk 12 above)
Comment (borisstoyanov) [reply test/integration/smoke/test_async_job.py]: I've tried this case and I'm not sure it baheves as expected. I've successfully destroyed a VM and tried to do it again and the job result was successful same as the first try. Should this behave like that? I think it should return an error saying the VM is already in Destroyed state.

File: test/integration/smoke/test_async_job.py (hunk 11 above)
Comment (borisstoyanov) [reply test/integration/smoke/test_async_job.py]: Honestly I'm not a fan of combining separate tests into one, this is not a good practice since it's limiting the test granularity. But in this case maybe it does makes sense to do it since the api result is depended on the DB value and could not work without it. In this case it'll save us the time to deploy a VM which is a lot and it does makes sense to combine.

File: framework/jobs/src/main/java/org/apache/cloudstack/framework/jobs/dao/AsyncJobDaoImpl.java (hunk 2 above)
Comment (ernjvr) [reply framework/jobs/src/main/java/org/apache/cloudstack/framework/jobs/dao/AsyncJobDaoImpl.java:74]: This is only used by the Asyncjob garbage collector.

File: framework/jobs/src/main/java/org/apache/cloudstack/framework/jobs/dao/AsyncJobDaoImpl.java (hunk 3 above)
Comment (ernjvr) [reply framework/jobs/src/main/java/org/apache/cloudstack/framework/jobs/dao/AsyncJobDaoImpl.java:173]: Update added to schema-41110to41200.sql

File: framework/jobs/src/main/java/org/apache/cloudstack/framework/jobs/dao/AsyncJobDaoImpl.java (hunk 2 above)
Comment (ernjvr) [reply framework/jobs/src/main/java/org/apache/cloudstack/framework/jobs/dao/AsyncJobDaoImpl.java:74]: I verified; it is only used by the async jobs garbage collector.

File: framework/jobs/src/main/java/org/apache/cloudstack/framework/jobs/dao/AsyncJobDaoImpl.java (hunk 3 above)
Comment (ernjvr) [reply framework/jobs/src/main/java/org/apache/cloudstack/framework/jobs/dao/AsyncJobDaoImpl.java:173]: Update added.

File: framework/jobs/src/main/java/org/apache/cloudstack/framework/jobs/dao/AsyncJobDaoImpl.java (hunk 9 above)
Comment (ernjvr) [reply framework/jobs/src/main/java/org/apache/cloudstack/framework/jobs/dao/AsyncJobDaoImpl.java]: reverted

File: test/integration/smoke/test_async_job.py (hunk 10 above)
Comment (ernjvr) [reply test/integration/smoke/test_async_job.py]: done

File: test/integration/smoke/test_async_job.py (hunk 11 above)
Comment (ernjvr) [reply test/integration/smoke/test_async_job.py]: done

File: test/integration/smoke/test_async_job.py (hunk 12 above)
Comment (ernjvr) [reply test/integration/smoke/test_async_job.py]: I get the same result as you @borisstoyanov

Comment (borisstoyanov) [reply test/integration/smoke/test_async_job.py]: @ernjvr do you think we should refactor the service if it tries to bring a VM into a state which it is, to prompt an error? It kinda makes sense to me.

Comment (ernjvr) [reply test/integration/smoke/test_async_job.py]: @borisstoyanov I investigated the code for the scenario of starting a VM. It also returns a success status when I make a request to start a VM that has already been started. This success status is returned by the VM job handler code. The async job status gets updated with whatever status is returned by the VM job handler code. So, I think changing how the VM handler code deals with duplicate commands for the same resource is out of scope for this feature. I think it should be dealt with in a separate bug fix ticket. What do you think?

Comment (borisstoyanov) [reply test/integration/smoke/test_async_job.py]: Sound the right way to do it, thanks for the investigation @ernjvr, let me log a ticket in community. When we have time we could get back to this.

Comment (borisstoyanov) [reply test/integration/smoke/test_async_job.py]: https://github.com/apache/cloudstack/issues/2736

File: test/integration/smoke/test_async_job.py [hunk 13]
```diff
@@ -0,0 +1,140 @@
...
+
+    def tearDown(self):
+        try:
+            self.debug("Cleaning up the resources")
+            cleanup_resources(self.apiclient, self.cleanup)
+            self.debug("Cleanup complete!")
+        except Exception as e:
+            self.debug("Warning! Exception in tearDown: %s" % e)
+
+    @attr(
```
Comment (yadvr) [inline test/integration/smoke/test_async_job.py]: This can be formatted to a single line.

File: test/integration/smoke/test_async_job.py [hunk 14]
```diff
@@ -0,0 +1,140 @@
...
+        tags=[
+            "advanced",
+            "eip",
+            "advancedns",
+            "basic",
+            "sg"],
+        required_hardware="true")
```
Comment (yadvr) [inline test/integration/smoke/test_async_job.py]: This could be set to required_hardware="false".

File: test/integration/smoke/test_async_job.py [hunk 15]
```diff
@@ -0,0 +1,140 @@
...
+            diskofferingid=self.disk_offering.id,
+            hypervisor=self.hypervisor
+        )
+
+        response = self.virtual_machine.getState(
+            self.apiclient,
+            VirtualMachine.RUNNING)
+        self.assertEqual(response[0], PASS, response[1])
+
+        result = self.dbclient.execute("select * from async_job where uuid='%s'" % self.virtual_machine.jobid)
```
Comment (yadvr) [inline test/integration/smoke/test_async_job.py]: Move this to line 127, closer to the access on `result[0][17]`, probably name this as db_result to make it more obvious.

File: test/integration/smoke/test_async_job.py [hunk 16]
```diff
@@ -0,0 +1,140 @@
...
+        cmd = queryAsyncJobResult.queryAsyncJobResultCmd()
+        cmd.jobid = self.virtual_machine.jobid
+        cmd_response = self.apiclient.queryAsyncJobResult(cmd)
+
+        # verify that 'completed' value from api equals 'removed' db column value
+        completed = cmd_response.completed
+        removed = timezone('UTC').localize(result[0][17])
+        removed = removed.astimezone(timezone('CET'))
+        removed = removed.strftime("%Y-%m-%dT%H:%M:%S%z")
+        self.assertEqual(completed, removed, "Expected 'completed' tag value to be equal to 'removed' db column value.")
```
Comment (yadvr) [inline test/integration/smoke/test_async_job.py]: Remove the word usage `tag` and use timestamp/datetime instead, it's good enough otherwise. Same at line 137.

**Review by yadvr — APPROVED** [review]
Left minor nits, otherwise LGTM. I did not test it, which the internal testing would approve or advise bugs.

File: test/integration/smoke/test_async_job.py [hunk 17]
```diff
@@ -0,0 +1,140 @@
...
+
+        # verify that api job_status value equals db job_status value
+        jobstatus_api = cmd_response.jobstatus
+        jobstatus_db = result[0][8]
+        self.assertEqual(jobstatus_api, jobstatus_db, "Expected 'jobstatus' tag value to be equal to 'job_status' db column value.")
+
+        return
```
Comment (yadvr) [inline test/integration/smoke/test_async_job.py]: You don't return at the end .

File: test/integration/smoke/test_async_job.py [hunk 18]
```diff
@@ -0,0 +1,140 @@
...
+        self.testdata["virtual_machine"]["zoneid"] = self.zone.id
+        self.testdata["virtual_machine"]["template"] = self.template.id
+        self.testdata["iso"]["zoneid"] = self.zone.id
+        self.account = Account.create(
+            self.apiclient,
+            self.testdata["account"],
+            domainid=self.domain.id
+        )
+        self.cleanup = [self.account]
+        return
```
Comment (yadvr) [inline test/integration/smoke/test_async_job.py]: Remove the `return`.

File: test/integration/smoke/test_async_job.py [hunk 19]
```diff
@@ -0,0 +1,140 @@
...
+# specific language governing permissions and limitations
+# under the License.
+
+from nose.plugins.attrib import attr
+from marvin.cloudstackTestCase import *
+from marvin.lib.utils import *
+from marvin.lib.base import *
+from marvin.lib.common import *
+from pytz import timezone
+
```
Comment (yadvr) [inline test/integration/smoke/test_async_job.py]: Run pep8 or pylint on this file and fix and lint related issues.
